Delay the loading icon display

calixteman opened the pull request:
In most of the cases, showing the loading icon is useless because
it's for a very short time, consequently it doesn't bring any useful
information for the user.
After a delay (400ms), the icon is shown in order to inform the user
that the viewer isn't stuck but it's doing something.

Snuffleupagus commented:
Sorry, but I don't really agree that this is a good/necessary change to make and it also adds a bunch of (what feels like unneeded) complexity.

If we do this it becomes *generally impossible* for the user to tell if a blank page means that there's actually no content on said page, or if the page has just not been loaded/rendered yet. To me this really doesn't feel improvement, but rather the opposite.

Snuffleupagus commented on web/app.js:
```diff
@@ -2331,8 +2333,21 @@ function webViewerUpdateViewarea({ location }) {
+  if (currentPage?.renderingState !== RenderingStates.FINISHED) {
+    PDFViewerApplication._loadIconTimeoutId = setTimeout(() => {
+      PDFViewerApplication._loadIconTimeoutId = null;
+      const loading = currentPage?.renderingState !== RenderingStates.FINISHED;
```
Also, this isn't even correct since there's nothing that says that `currentPage` is actually still *the current page* when the timeout resolves.

calixteman commented:
> Sorry, but I don't really agree that this is a good/necessary change to make and it also adds a bunch of (what feels like unneeded) complexity.
> 
> If we do this it becomes _generally impossible_ for the user to tell if a blank page means that there's actually no content on said page, or if the page has just not been loaded/rendered yet. To me this really doesn't feel like an improvement, but rather the opposite.

If the page is really blank, the icon will be displayed for 1 or 2 ms, maybe it'll never been displayed.
Especially if the user is scrolling, then they'll never see it on a "normal" page because the page will be likely rendered before they reach the page center.
I don't really see how it's useful for the user to have this icon during few hundred of millis, but I can change my mind.
Note that I don't want to remove the icon since I really think it's useful, but under a certain delay it's really useless.
I chose 1500ms but we can change this value: I've no strong opinions about a good value here.

calixteman commented:
That said, I really rely on your judgement: you really know this project better than me.
So if it's a strong no, just tell me and I'll remove this PR.
If you're so-so, I can ask to UI/UX what they think about that, but as far as I can tell, in general we want to try to avoid useless stuff which wastes resources like power.

Snuffleupagus commented:
> If you're so-so, I can ask to UI/UX what they think about that,

Let's start there, please. (Since UI design isn't exactly my strong suite.)

Snuffleupagus commented:
Implementation-wise, the timeout usage is what I'm the least "excited" about here since that'll lead to a bunch of complexity that doesn't currently exist.
So how about if we only display the loadingIcon-div on the page which currently has `RenderingState.RUNNING` (without any timeouts), i.e. the most visible page, since that would actually allow us some overall simplification already; my take on this in https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:loadingIcon-toggle

calixteman commented:
> Implementation-wise, the timeout usage is what I'm the least "excited" about here since that'll lead to a bunch of complexity that doesn't currently exist. So how about if we only display the loadingIcon-div on the page which currently has `RenderingState.RUNNING` (without any timeouts), i.e. the most visible page, since that would actually allow us some overall simplification already; my take on this in [master...Snuffleupagus:pdf.js:loadingIcon-toggle](https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:loadingIcon-toggle)

Clearly what you're proposing will simplify this stuff: so +1 for me.
I understand you aren't excited by the timeout stuff: I'm not either. I asked to UX but I've no answer for the moment (a lot of people are in pto right now), but as far as I can tell, this is something we'll do in other parts of firefox.
Anyway, what about using https://developer.mozilla.org/en-US/docs/Web/CSS/transition-delay ? which should do exactly what we want here.

calixteman committed:
Delay the loading icon display

In most of the cases, showing the loading icon is useless because
it's for a very short time, consequently it doesn't bring any useful
information for the user.
After a delay (400ms), the icon is shown in order to inform the user
that the viewer isn't stuck but it's doing something.

Snuffleupagus commented on web/pdf_viewer.css:
```diff
@@ -157,10 +157,13 @@
   right: 0;
   bottom: 0;
   background: url("images/loading-icon.gif") center no-repeat;
+  visibility: visible;
+  transition-property: visibility;
```
Any reason that we couldn't just use the existing `background`-property instead, rather than introducing `visibility` here?

In any case, a good comment explaining what's happening here seems necessary to aid future readers (that might even be us in a few months time).

calixteman replied:
I tried different things:
 - transition on `background` leads to finally transition on the position and not the image.
 - transition on `background-image` doesn't work.

We could make the transition on background-size (like for the other one), but in this case making the transition on the visibility avoid to have to know the dimensions and I thought it makes senses to change the visibility since we're toggling `notVisible`.

Snuffleupagus commented on web/viewer.css:
```diff
@@ -1104,11 +1104,15 @@ a.secondaryToolbarButton[href="#"] {
   -moz-appearance: textfield; /* hides the spinner in moz */
   text-align: right;
   width: 40px;
+  background-size: 0 0;
+  transition-property: background-size;
```
Similar to the comment above, could we not use the existing `background-image`-property instead, since using `background-size` feels like a hack?

In any case, a good comment explaining what's happening here seems necessary to aid future readers (that might even be us in a few months time).

calixteman replied:
Making the transition on `background-image` just doesn't work in Firefox. We can't change the visibility as above because this is the background of an `input` so the only workaround I found is to make the transition on `background-size`.

Snuffleupagus reviewed (commented):
I think that it makes sense to wait for UX feedback before moving forward here, and in any case we probably don't want to land it just before a release.

Also, based on reading the patch I've got one "larger" question: Will this apply to both showing *and* hiding of the loadingIcons? Because if so, that's probably not what we want here since the hiding needs to be immediate.

calixteman commented:
400ms is the value we're using for the tab spinners, so use the same.

calixteman commented:
UI/UX agreed on 400ms.

Snuffleupagus commented on web/pdf_viewer.css:
```diff
@@ -25,6 +25,7 @@
   --page-border-image: url(images/shadow.png) 9 9 repeat;
   --spreadHorizontalWrapped-margin-LR: -3.5px;
   --scale-factor: 1;
+  --loading-delay: 400ms;
```
How about the following name instead, to make it slightly more clear what it's being used for?
```suggestion
  --loading-icon-delay: 400ms;
```

Snuffleupagus commented on web/pdf_viewer.css:
```diff
@@ -157,10 +158,16 @@
   bottom: 0;
   background: url("images/loading-icon.gif") center no-repeat;
+  visibility: visible;
+  /* Using a delay with bacground-image doesn't work,
+     consequently, we use the visibility. */
```
Spelling/grammar nits:
```suggestion
  /* Using a delay with background-image doesn't work,
     consequently we use the visibility. */
```

Snuffleupagus commented on web/viewer.css:
```diff
@@ -1104,11 +1104,18 @@ a.secondaryToolbarButton[href="#"] {
   background-repeat: no-repeat;
   background-position: 3px;
+  background-size: 16px 16px;
+  /* Using a delay with bacground-image doesn't work,
```
Spelling nit:
```suggestion
  /* Using a delay with background-image doesn't work,
```

Snuffleupagus reviewed (approved):
r=me, with comments addressed, since this seems to work in my testing (and a delay of only 400ms isn't so long that it renders the indicators pointless).

calixteman commented:
/botio integrationtest
